spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/PushDown.scala

```diff
@@ -0,0 +1,204 @@
+    val inColumns = otherColumns.headOption.toSeq.filter(inPredicatesByName.contains)
+    if (eqColumns.size + inColumns.size == partitionKeyColumns.size)
+      eqColumns.flatMap(eqPredicatesByName) ++ inColumns.flatMap(inPredicatesByName)
+    else
+      Nil
+  }
+
+  /**
+   * Selects clustering key predicates for pushdown:
+   * 1. Clustering column predicates must be equality predicates, except the last one.
+   * 2. The last predicate is allowed to be an equality or a range predicate.
+   * 3. The last predicate is allowed to be an IN predicate only if it was preceded by an equality predicate.
+   * 4. Consecutive clustering columns must be used, but, contrary to partition key, the tail can be skipped.
+   */
+  private val clusteringColumnPredicatesToPushDown: Seq[T] = {
+    val (eqColumns, otherColumns) = clusteringColumns.span(eqPredicatesByName.contains)
+    val eqPredicates = eqColumns.flatMap(eqPredicatesByName)
+    val optionalNonEqPredicate = for {
+      c <- otherColumns.headOption.toSeq
+      p <- firstNonEmptySeq(rangePredicatesByName(c), inPredicatesByName(c).filter(
+        _ => c==clusteringColumns.last))
+    } yield p
+
+    eqPredicates ++ optionalNonEqPredicate
+  }
+
+  /**
+   * Selects indexed and regular column predicates for pushdown:
+   * 1. At least one indexed column must be present in an equality predicate to be pushed down.
+   * 2. Regular column predicates can be either equality or range predicates.
+   * 3. If multiple predicates use the same column, equality predicates are preferred over range predicates.
+   */
+  private val indexedColumnPredicatesToPushDown: Seq[T] = {
+    val inPredicateInPrimaryKey = partitionKeyPredicatesToPushDown.exists(isIn)
+    val eqIndexedColumns = indexedColumns.filter(eqPredicatesByName.contains)
+    val eqIndexedPredicates = eqIndexedColumns.flatMap(eqPredicatesByName)
+    // Don't include partition predicates as None-indexed predicates if partition predicates can't
+    // be pushed down because we use token range query which already has partition columns in the
+    // where clause and it can't have other partial partition columns in where clause any more.
+    val nonIndexedPredicates = for {
+      c <- allColumns if !partitionKeyPredicatesToPushDown.isEmpty && !eqIndexedColumns.contains(c) ||
+      partitionKeyPredicatesToPushDown.isEmpty && !eqIndexedColumns.contains(c) &&
+        !partitionKeyColumns.contains(c)
+      p <- firstNonEmptySeq(eqPredicatesByName(c), rangePredicatesByName(c))
+    } yield p
+
+    if (!inPredicateInPrimaryKey && eqIndexedColumns.nonEmpty)
+      eqIndexedPredicates ++ nonIndexedPredicates
+    else
+      Nil
+  }
+
+  val toPushDown = (
+    partitionKeyPredicatesToPushDown ++
+      clusteringColumnPredicatesToPushDown ++
+      indexedColumnPredicatesToPushDown).distinct
+
+  val toPreserve = predicates.filterNot(toPushDown.toSet.contains)
```

Comment:
If the order doesn't matter, you can do just `predicates.toSet -- toPushDown.toSet`